File: bindings/dart/lib/src/netconf.dart
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// NETCONF protocol types for proven-servers.

/// NetconfOperation matching the Idris2 ABI tags.
enum NetconfOperation {
  get_(0),
  getConfig(1),
  editConfig(2),
  copyConfig(3),
  deleteConfig(4),
  lock(5),
  unlock(6),
  closeSession(7),
  killSession(8),
  commit(9),
  validate(10),
  discardChanges(11);

  const NetconfOperation(this.tag);
  final int tag;

  static NetconfOperation? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// Datastore matching the Idris2 ABI tags.
enum Datastore {
  running(0),
  startup(1),
  candidate(2);

  const Datastore(this.tag);
  final int tag;

  static Datastore? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// EditOperation matching the Idris2 ABI tags.
enum EditOperation {
  merge(0),
  replace(1),
  create(2),
  delete(3),
  remove(4);

  const EditOperation(this.tag);
  final int tag;

  static EditOperation? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// NetconfErrorType matching the Idris2 ABI tags.
enum NetconfErrorType {
  transport(0),
  rpc(1),
  protocol(2),
  application(3);

  const NetconfErrorType(this.tag);
  final int tag;

  static NetconfErrorType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ErrorSeverity matching the Idris2 ABI tags.
enum ErrorSeverity {
  error(0),
  warning(1);

  const ErrorSeverity(this.tag);
  final int tag;

  static ErrorSeverity? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// NetconfState matching the Idris2 ABI tags.
enum NetconfState {
  idle(0),
  connected(1),
  locked(2),
  editing(3),
  closing(4),
  terminated(5);

  const NetconfState(this.tag);
  final int tag;

  static NetconfState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

File: bindings/javascript/src/netconf.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// NETCONF protocol types for proven-servers.

/** NetconfOperation matching the Idris2 ABI tags. */
export const NetconfOperation = Object.freeze({
  GET: 0,
  GET_CONFIG: 1,
  EDIT_CONFIG: 2,
  COPY_CONFIG: 3,
  DELETE_CONFIG: 4,
  LOCK: 5,
  UNLOCK: 6,
  CLOSE_SESSION: 7,
  KILL_SESSION: 8,
  COMMIT: 9,
  VALIDATE: 10,
  DISCARD_CHANGES: 11,
});

/** Datastore matching the Idris2 ABI tags. */
export const Datastore = Object.freeze({
  RUNNING: 0,
  STARTUP: 1,
  CANDIDATE: 2,
});

/** EditOperation matching the Idris2 ABI tags. */
export const EditOperation = Object.freeze({
  MERGE: 0,
  REPLACE: 1,
  CREATE: 2,
  DELETE: 3,
  REMOVE: 4,
});

/** NetconfErrorType matching the Idris2 ABI tags. */
export const NetconfErrorType = Object.freeze({
  TRANSPORT: 0,
  RPC: 1,
  PROTOCOL: 2,
  APPLICATION: 3,
});

/** ErrorSeverity matching the Idris2 ABI tags. */
export const ErrorSeverity = Object.freeze({
  ERROR: 0,
  WARNING: 1,
});

/** NetconfState matching the Idris2 ABI tags. */
export const NetconfState = Object.freeze({
  IDLE: 0,
  CONNECTED: 1,
  LOCKED: 2,
  EDITING: 3,
  CLOSING: 4,
  TERMINATED: 5,
});
